src/types.h

```diff
@@ -42,50 +44,80 @@ typedef struct {
+    STATE_NONE,      /// No state
+    STATE_RECV_DATA, /// Receiving data (for SIGN_TX)
+    STATE_PARSED,    /// Transaction data parsed
+    STATE_APPROVED   /// Transaction data approved
 } state_e;
 
 /**
  * Enumeration with user request type.
  */
 typedef enum {
     CONFIRM_ADDRESS,     /// confirm address derived from public key
+    CONFIRM_XPUB,        /// confirm access to xpub
     CONFIRM_TRANSACTION  /// confirm transaction information
 } request_type_e;
 
+
+#define MAX_SCREEN_LENGTH 12
 /**
  * Structure for public key context information.
  */
 typedef struct {
-    uint8_t raw_public_key[64];  /// x-coordinate (32), y-coodinate (32)
+    bip32_path_t key_path;
+
+    uint8_t raw_public_key[65];
     uint8_t chain_code[32];      /// for public key derivation
-} pubkey_ctx_t;
+    uint8_t fingerprint[4];
+} xpub_ctx_t;
 
 /**
  * Structure for transaction information context.
  */
 typedef struct {
-    uint8_t raw_tx[MAX_TRANSACTION_LEN];  /// raw transaction serialized
-    size_t raw_tx_len;                    /// length of raw transaction
-    transaction_t transaction;            /// structured transaction
-    uint8_t m_hash[32];                   /// message hash digest
-    uint8_t signature[MAX_DER_SIG_LEN];   /// transaction signature encoded in DER
-    uint8_t signature_len;                /// length of transaction signature
-    uint8_t v;                            /// parity of y-coordinate of R in ECDSA signature
-} transaction_ctx_t;
+    uint8_t buffer[300];
+    size_t buffer_len;
+    // sha256 context for the hash
+    cx_sha256_t sha256;
+    uint8_t sighash_all[32];
+
+    bool has_change_output;
+    uint8_t change_output_index;
+    // uint8_t change_bip32_len;
+    // uint32_t change_bip32_path[MAX_BIP32_PATH];
+    bip32_path_t change_bip32_path;
+    // tx
+    uint16_t tx_version;
+    uint8_t remaining_tokens;
+    uint8_t remaining_inputs;
+    uint8_t outputs_len;
+
+    // type of decoded element
+    uint8_t elem_type;
+    uint8_t current_output;
+    tx_output_t decoded_output;
+
+    // display value
+    unsigned char info[70]; // address + HTR value
+    uint8_t display_index;
+    // NULL terminated string for display
+    char line1[15];
+    char line2[12 + 1]; // MAX_SCREEN_LEN(12) + 1
+
+} sign_tx_ctx_t;
 
 /**
  * Structure for global context.
  */
 typedef struct {
     state_e state;  /// state of the context
     union {
-        pubkey_ctx_t pk_info;       /// public key context
-        transaction_ctx_t tx_info;  /// transaction context
+        xpub_ctx_t pk_info;       /// public key context
+        sign_tx_ctx_t tx_info;  /// transaction context
     };
     request_type_e req_type;              /// user request
-    uint32_t bip32_path[MAX_BIP32_PATH];  /// BIP32 path
-    uint8_t bip32_path_len;               /// lenght of BIP32 path
+    bip32_path_t bip32_path;
+    // uint32_t bip32_path[MAX_BIP32_PATH];  /// BIP32 path
```

Comment:
It seems we are not using these two commented lines anymore. Can we remove them?

Reply:
Done.